charm/src/charm.py

```diff
@@ -0,0 +1,186 @@
+
+    def _update_layer_and_restart(self, event) -> None:
+        """Define and start layer for testflinger using Pebble"""
+
+        if not self.container.can_connect():
+            self.unit.status = ops.WaitingStatus(
+                "Waiting for Pebble in workload container"
+            )
+            return
+
+        self.unit.status = ops.MaintenanceStatus("Assembling pod spec")
+        new_layer = self._pebble_layer.to_dict()
+
+        # Get the current pebble layer config
+        services = self.container.get_plan().to_dict().get("services", {})
+        if services != new_layer["services"]:
+            # Changes were made, add the new layer
+            self.container.add_layer(
+                "testflinger", self._pebble_layer, combine=True
+            )
+            logger.info("Added updated layer 'testflinger' to Pebble plan")
+
+            self.container.restart(self.pebble_service_name)
+            logger.info("Restarted '%s' service", self.pebble_service_name)
+
+        # add workload version in juju status
+        self.unit.set_workload_version(self.version)
+        self.unit.status = ops.ActiveStatus()
+
+    def _on_mongodb_client_relation_removed(
+        self, event: ops.framework.EventBase
+    ) -> None:
+        """Event is fired when relation with mongodb is broken."""
+        self.unit.status = ops.WaitingStatus("Waiting for database relation")
+        raise SystemExit(0)
+
+    def _on_config_changed(self, event: ops.framework.EventBase) -> None:
+        """Handle config changed event"""
+        self._update_layer_and_restart(event)
+
+    @property
+    def _pebble_layer(self):
+        """Return a dictionary representing a Pebble layer."""
+        command = " ".join(
+            [
+                "gunicorn",
+                "--bind",
+                "0.0.0.0:5000",
+                "testflinger:app",
+            ]
+        )
+        pebble_layer = {
+            "summary": "Testflinger server",
+            "description": "pebble config layer for Testflinger server",
+            "services": {
+                self.pebble_service_name: {
+                    "override": "replace",
+                    "summary": "testflinger",
+                    "command": command,
+                    "startup": "enabled",
+                    "environment": self.app_environment,
+                }
+            },
+        }
+        return Layer(pebble_layer)
+
+    @property
+    def app_environment(self) -> dict:
+        """Get dict of env data for the mongodb credentials"""
+        db_data = self.fetch_mongodb_relation_data()
+        env = {
+            "MONGODB_HOST": db_data.get("db_host", None),
+            "MONGODB_PORT": db_data.get("db_port", None),
+            "MONGODB_USERNAME": db_data.get("db_username", None),
+            "MONGODB_PASSWORD": db_data.get("db_password", None),
+            "MONGODB_DATABASE": db_data.get("db_database", None),
```

Comment:
It’s None by default, so you don't need to specify it explicitly

Reply:
done